File: src/services/webwin.ts
// Open (or focus) a dedicated persistent window for a web app with no usable API (WhatsApp, LinkedIn).
// Backed by the Rust open_web_window command; the session persists so you log in once.
export async function openWebWindow(label: string, url: string, title: string): Promise<void> {
  const { invoke } = await import('@tauri-apps/api/core')
  await invoke('open_web_window', { label, url, title })
}

export const WEB_APPS: Record<string, { label: string; url: string; title: string; glyph: string }> = {
  whatsapp: { label: 'whatsapp', url: 'https://web.whatsapp.com',  title: 'WhatsApp', glyph: '◍' },
  linkedin: { label: 'linkedin', url: 'https://www.linkedin.com/feed/', title: 'LinkedIn', glyph: 'in' },
}

File: src-tauri/src/webwin.rs
use tauri::{Manager, WebviewUrl, WebviewWindowBuilder};

// Open (or focus) a dedicated window pointed at an external web app — used for services with no
// usable API (WhatsApp Web, LinkedIn). The webview persists cookies/storage in the app's data dir,
// so the user scans the WhatsApp QR / logs in once and stays signed in across opens.
#[tauri::command]
pub fn open_web_window(
    app: tauri::AppHandle,
    label: String,
    url: String,
    title: String,
    width: Option<f64>,
    height: Option<f64>,
) -> Result<(), String> {
    if let Some(existing) = app.get_webview_window(&label) {
        let _ = existing.show();
        let _ = existing.set_focus();
        return Ok(());
    }
    let parsed: tauri::Url = url.parse().map_err(|_| format!("invalid url: {url}"))?;
    WebviewWindowBuilder::new(&app, &label, WebviewUrl::External(parsed))
        .title(&title)
        .inner_size(width.unwrap_or(1040.0), height.unwrap_or(780.0))
        .build()
        .map_err(|e| e.to_string())?;
    Ok(())
}
